ads/google/a4a/google-data-reporter.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {EXPERIMENT_ATTRIBUTE, getCorrelator, QQID_HEADER} from './utils';
+import {BaseLifecycleReporter, GoogleAdLifecycleReporter} from './performance';
+import {getMode} from '../../../src/mode';
+import {isExperimentOn} from '../../../src/experiments';
+
+import {
+    parseExperimentIds,
+    isInManualExperiment,
+    randomlySelectUnsetPageExperiments,
+} from './traffic-experiments';
+import {
+    ADSENSE_A4A_EXTERNAL_EXPERIMENT_BRANCHES,
+    ADSENSE_A4A_INTERNAL_EXPERIMENT_BRANCHES,
+} from '../../../extensions/amp-ad-network-adsense-impl/0.1/adsense-a4a-config';
+import {
+    DOUBLECLICK_A4A_EXTERNAL_EXPERIMENT_BRANCHES,
+    DOUBLECLICK_A4A_INTERNAL_EXPERIMENT_BRANCHES,
+} from '../../../extensions/amp-ad-network-doubleclick-impl/0.1/doubleclick-a4a-config';  // eslint-disable-line max-len
+
+/**
+ * Check whether the element is in an experiment branch that is eligible for
+ * monitoring.
+ *
+ * @param {!AMP.BaseElement} ampElement
+ * @param {!string} namespace
+ * @returns {boolean}
+ */
+function isInReportableBranch(ampElement, namespace) {
+  // Handle the possibility of multiple eids on the element.
+  const eids = parseExperimentIds(
+      ampElement.element.getAttribute(EXPERIMENT_ATTRIBUTE));
+  const reportableA4AEids = {
+    [ADSENSE_A4A_EXTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+    [ADSENSE_A4A_INTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+    [DOUBLECLICK_A4A_EXTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+    [DOUBLECLICK_A4A_INTERNAL_EXPERIMENT_BRANCHES.experiment]: 1,
+  };
+  const reportableControlEids = {
+    [ADSENSE_A4A_EXTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+    [ADSENSE_A4A_INTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+    [DOUBLECLICK_A4A_EXTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+    [DOUBLECLICK_A4A_INTERNAL_EXPERIMENT_BRANCHES.control]: 1,
+  };
+  if (namespace == 'a4a' &&
```

Comment:
Couple of thoughts here:

- Could we make namespace an enum?
- Code would likely be easier to read if we use a switch on namespace.  Something like:
`switch (namespace) {
   case 'a4a':
       return (eids.some(x => { return x in reportableA4AEids; }) || isInManualExperiment(ampElement.element));
   case 'amp':
       return eids.some(x => { return x in reportableControlEids; });
   default:
       return false;
}`